resolwe_bio/processes/clustering/pca.py

```diff
@@ -0,0 +1,334 @@
+    - log2: use log2(x+const) transformation
+    - const: an additive constant used in computation of log2
+    - z_score: use Z-score normalization using StandardScaler
+    """
+    if log2:
+        expressions = expressions.applymap(lambda x: np.log2(x + const))
+        if expressions.isnull().values.any():
+            error(
+                "Cannot apply log2 to expression values. Some expression values are negative."
+            )
+    if z_score:
+        try:
+            scaler = StandardScaler()
+            expressions = expressions.transpose()
+            expressions_arr = scaler.fit_transform(expressions)
+            expressions = pd.DataFrame(
+                expressions_arr, index=expressions.index, columns=expressions.columns
+            ).transpose()
+        except ValueError:
+            error("Cannot apply Z-score normalization to expression values.")
+
+    return expressions
+
+
+def get_csv(fnames):
+    """Read CSV files, perform transformations and return Pandas DataFrame."""
+    expressions = [read_csv(fname) for fname in fnames]
+    return pd.concat(expressions, axis=1, join="inner")
+
+
+def run_pca(
+    exp_paths,
+    rc_paths,
+    sample_ids,
+    gene_labels,
+    log2_transform,
+    standard_scaler,
+    error,
+):
+    """Read data, run PCA, and output results."""
+    expressions = get_csv(exp_paths)
+    # Filter out genes with low expression values
+    if rc_paths:
+        read_counts = get_csv(rc_paths)
+        percentile_80 = np.percentile(read_counts, 80, axis=1)
+        retained_genes = percentile_80 > 100
+        expressions = expressions.loc[retained_genes]
+
+    if expressions.empty:
+        error(
+            "Gene selection and filtering resulted in no genes. Please select different samples or genes."
+        )
+
+    expressions = transform(
+        error=error,
+        expressions=expressions,
+        log2=log2_transform,
+        z_score=standard_scaler,
+    )
+
+    if gene_labels:
+        gene_labels = list(set(gene_labels).intersection(expressions.index))
+        gene_labels.sort()
+        expressions = expressions.loc[gene_labels]
+
+    result = get_pca(expressions, gene_labels, error)
+    save_pca(result, sample_ids)
+
+
+class PCA(Process):
+    """Principal component analysis process."""
+
+    slug = "pca-beta"
+    name = "PCA"
+    requirements = {
+        "expression-engine": "jinja",
+        "executor": {
+            "docker": {
+                "image": "public.ecr.aws/genialis/resolwebio/rnaseq:6.0.0",
+            },
+        },
+        "resources": {
+            "cores": 1,
+            "memory": 4096,
+            "storage": 10,
+        },
+    }
+    data_name = "{{ alignment|name|default('?') }}"
+    version = "1.0.0"
+    process_type = "data:pca"
+    category = "Enrichment and Clustering"
+    scheduling_class = SchedulingClass.INTERACTIVE
+    persistence = Persistence.TEMP
+
+    description = "Principal component analysis (PCA)"
```

Comment:
If we decide to infer species and source internally, it could be specified here that we are assuming they all match.